src/ray/gcs/gcs_client/gcs_client.cc

```diff
@@ -260,90 +266,146 @@ bool GcsClient::GetGcsServerAddressFromRedis(redisContext *context,
-  if (get_server_address_func_(&address)) {
-    if (address != current_gcs_server_address_) {
-      // If GCS server address has changed, invoke the `GcsServiceFailureDetected`
-      // callback.
+  if (!get_server_address_func_(&address)) {
+    return;
+  }
+
+  if (address != current_gcs_server_address_) {
+    // If GCS server address has changed, invoke the `GcsServiceFailureDetected`
+    // callback.
+    // NOTE(wanxing): Only update `current_gcs_server_address_` if reconnection is
+    // accepted.
+    if (GcsServiceFailureDetected(rpc::GcsServiceFailureType::GCS_SERVER_RESTART)) {
       current_gcs_server_address_ = address;
-      GcsServiceFailureDetected(rpc::GcsServiceFailureType::GCS_SERVER_RESTART);
     }
   }
 }
 
-void GcsClient::GcsServiceFailureDetected(rpc::GcsServiceFailureType type) {
+bool GcsClient::GcsServiceFailureDetected(
+    rpc::GcsServiceFailureType type,
+    const std::function<void()> failure_handled_callback) {
+  RAY_LOG(INFO) << "Gcs service failure detected, reason "
+                << GcsServiceFailureType_Name(type);
+  bool accepted = false;
   switch (type) {
   case rpc::GcsServiceFailureType::RPC_DISCONNECT:
     // If the GCS server address does not change, reconnect to GCS server.
-    ReconnectGcsServer();
+    accepted = ReconnectGcsServerAsync(failure_handled_callback);
     break;
   case rpc::GcsServiceFailureType::GCS_SERVER_RESTART:
     // If GCS sever address has changed, reconnect to GCS server and redo
     // subscription.
-    ReconnectGcsServer();
-    // If using GCS server for pubsub, resubscribe to GCS publishers.
-    resubscribe_func_(RayConfig::instance().gcs_grpc_based_pubsub());
-    // Resend resource usage after reconnected, needed by resource view in GCS.
-    node_resource_accessor_->AsyncReReportResourceUsage();
+    accepted = ReconnectGcsServerAsync([this]() {
+      // If using GCS server for pubsub, resubscribe to GCS publishers.
+      resubscribe_func_(RayConfig::instance().gcs_grpc_based_pubsub());
+      // Resend resource usage after reconnected, needed by resource view in GCS.
+      node_resource_accessor_->AsyncReReportResourceUsage();
+    });
     break;
   default:
     RAY_LOG(FATAL) << "Unsupported failure type: " << type;
     break;
   }
+
+  return accepted;
 }
 
-void GcsClient::ReconnectGcsServer() {
+bool GcsClient::ReconnectGcsServerAsync(
+    const std::function<void()> reconnected_callback) {
+  absl::MutexLock l(&reconnect_flag_mutex_);
+  if (disconnected_) {
+    return false;
+  }
+  {
+    absl::MutexLock l(&reconnect_callbacks_mutex_);
+    reconnected_callbacks_.emplace_back(reconnected_callback);
+  }
+  if (reconnect_in_progress_) {
+    RAY_LOG(DEBUG) << "GcsClient reconnect in progress, the callback is pended.";
+    return false;
+  }
+  reconnect_in_progress_ = true;
+  DoReconnect(absl::Now());
+  return true;
+}
+
+void GcsClient::DoReconnect(absl::Time start) {
+  if (disconnected_) {
+    ExecuteReconnectedCallbacks();
```

Comment:
This is a bit weird when reading code. The signature said

```
  /// \param callback The callback function to be called after reconnected to gcs
  /// successfully.
```

But in this case, reconnect wasn't successful, but callback is still called.

Reply:
Hmm, The callbacks list should be cleared when disconnected without executing.